### PR TITLE
KPI J updates

### PR DESCRIPTION
**Pull request by tomkralidis**
Discussions with @gaubert indicate that this KPI is ready for implementation. I've made some editorial updates as well as added numeric scoring. For review and comment.

Note that scoring is firm/strict against bad links, and rewarding for accessible links.  IMHO bad links are a showstopper, hence the scoring.  Happy to hear your feedback and we can update accordingly

### CODE REVIEW

File: KPI/J.adoc
Comment (martaguti) [inline KPI/J.adoc:43]: Does the number of links make it a factor for the score grading.
Your record may have one link only which is accessible , does it get a 5 because there is only one. 
Maybe the scoring can be assessed against the ratio depending on how many links the record has 1/1 (10 points)
2/4 (5 points)

**Review by martaguti — COMMENTED** [review]
just commenting to clarify the scoring

File: KPI/J.adoc
Comment (tomkralidis) [reply KPI/J.adoc:43]: Good points @martaguti.  How about the following:

- Document has 1 or more broken links: 0
- Document has 1 - 3 accessible links: 2
- Document has 4 - 6 accessible links: 5
- Document has over 6 accessible links: 10

The idea being that even one broken link results in a 0.  Does this seem reasonable?

Comment (amilan17) [reply KPI/J.adoc:43]: I don't think this is reasonable. I don't think the score should drop to 0 if there is one bad link out of 10. Can we make this a percentage instead? 9 good out of 10 == 90%? 10 bad out of 10 == 0%?

Comment (tomkralidis) [reply KPI/J.adoc:43]: Good point. So do we measure number of links or percentage, or both?  Example:

- all links fail: 0
- 1 - 3 links with >= 75% pass: 2
- 4 - 6 links with >= 75% pass: 5
- 6 or more links with >= 75% pass: 10

Comment (amilan17) [reply KPI/J.adoc:43]: I don't think we should measure the count of links and the accessibility of links in one KPI. These are two different things. It's not necessarily better if there are 10 working links instead of 2 working links.

Comment (martaguti) [reply KPI/J.adoc:43]: I think I am with @amilan17  here the having more accessible links should not be a higher score factor. 
Percentage or ratio should work  IMHO 1 good out of 1 should score the same as 10 good out of 10. Otherwise people may start to put non necessary links just to get higher scores :-)

Comment (tomkralidis) [reply KPI/J.adoc:43]: Fair enough :)  So then what about pure percentage (i.e. number of good links as a percentage of total links)?  Else, happy to hear alternate proposals.

**Comment by tomkralidis** [comment]
Update: I've added HTTPS criteria to this KPI as per #66 and have subsequently named this KPI "Links health" as a result.

**Comment by gaubert** [comment]
We should also add the SSL certificate trustworthiness in mix. A link without a trusted certificate is like a broken link these days.
Browsers and command line tools like curl/wget will refuse by default to access the content of a web site with an invalid certificate.

**Comment by tomkralidis** [comment]
Makes sense @gaubert.  I've updated this PR with this precision.

**Comment by josusky** [comment]
WMO web just put me in a deep depression. I have tried to access a non existent link "https://www.wmo.int/pages/prog/wis/2012/metadata/version_1-4/". The server have responded with code "200 OK" and following text:
> <p>file does not exist: 2012</p
What a shame, dear WMO. It should have been code "404".

File: KPI/J.adoc
Comment (tomkralidis) [reply KPI/J.adoc:54]: FYI PR updated; I have added a relative scoring here based on number of links (total valid links / total links), for review and comment.

Comment (josusky) [reply KPI/J.adoc:54]: Thanks, Tom. I am having troubles to navigate in the threaded discussion about this PR and the 5 (?) related commits, sorry. I am fine with the score calculated as "total valid links / total links (100%)" but I am not sure what is the final decision about HTTPS. Are HTTP links considered invalid? BTW, the check I am adding to pywcmp can correctly recognize links starting with "http://" that, on actual access, are upgraded to HTTPS.
Or are we going to give only 80% for the link validity and reserve the additional 20% as a bonus for HTTPS use? A third option would be to have link validity and HTTPS use as two separate tests.

Comment (tomkralidis) [reply KPI/J.adoc:54]: Thanks @josusky.  What about two separate cases foreach link?
- link resolves
- link is HTTPS

So then each link would have a maximum of two points per above, and the final score of the KPI is a percentage?  So two valid non-HTTPS links would result in 2/4, or 50%.  Two valid HTTPS links would result in 4/4, of 100%.  Does this work?  If so, I will update this KPI.

For HTTP links that, on actual access, are upgraded to HTTPS, I would not count those as HTTPS per se, as they should be represented as HTTPS from the metadata document proper.  If the HTTP links redirect to HTTP on actual access, then the metadata editor would need to update their record accordingly.

**Comment by tomkralidis** [comment]
@gaubert / @josusky / @martaguti / @amilan17.  Updated as per today's discussion/actions.  Please review/comment and approve (if ready).

**Review by amilan17 — APPROVED** [review]
looks good

**Review by josusky — APPROVED** [review]

**Comment by gaubert** [comment]
looks fine +1

**Review by gaubert — APPROVED** [review]

File: KPI/J.adoc
Comment (martaguti) [inline KPI/J.adoc:41]: The codeList url is a copy of the gmxCodelists in WMO site 

Once we confirm which url are authoritatives we should cross check all codelists URL through all KPI's and give consistent examples.

**Pull request merged**
